### ISSUE TITLE
Pending/TTL Checkers use the same transaction

### ISSUE THREAD

**Issue by Zelldon** [camunda/zeebe#12647]
**Describe the bug**

On investigating https://github.com/camunda/zeebe/issues/12036#issuecomment-1532727109 I realized that our current checkers, like PendingDeployment/DeploymentRedistributor, MessageTTL, Timer due checker etc. are using the same transaction as the StreamProcessor on processing.

This is problematic because when processing a command we change the state, which is collected in the current transaction. These state changes are transient, they are not committed yet. At the point of processing and even after the transaction might be committed but could also be rolled back on error. It is not safe for "external" checkers to use the same transaction because there might be invalid assumptions done.

The **impact** is rather limited since the checkers shouldn't change the state (we know some do, which is bad) They might write invalid commands to the log, based on false assumptions which need to be rejected then on processing.

**To Reproduce**

- https://github.com/camunda/zeebe/blob/zell-set-prefix-filter/engine/src/main/java/io/camunda/zeebe/engine/state/ProcessingDbState.java#L110
- [Timer State ](https://github.com/camunda/zeebe/blob/zell-set-prefix-filter/engine/src/main/java/io/camunda/zeebe/engine/state/ProcessingDbState.java#L94) used by [DueChecker](https://github.com/camunda/zeebe/blob/zell-set-prefix-filter/engine/src/main/java/io/camunda/zeebe/engine/processing/EngineProcessors.java#L85)
- https://github.com/camunda/zeebe/blob/zell-set-prefix-filter/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobEventProcessors.java#L41
- etc.

**Expected behavior**

All "external" checkers, which are not inside the processing should use a different transaction. Ideally I would like to see a complete separation of the state classes, checkers and processors. Right now they are too coupled and it is not clear when and in which context which is used or called, which makes hard to reason about the code.

**Environment:**

- Zeebe Version: all

**Comment by koevskinikola** [comment]
ZPA triage:
- @Zelldon is this issue related to #12308 which @oleschoenburg is working on? Is this something that will be worked on in that scope?
- Note: for MessageTTL and Timer dueDate, the describtion above isn't true anymore, as they already run in a separate transaction (see [this](https://github.com/camunda/zeebe/blob/zell-set-prefix-filter/engine/src/main/java/io/camunda/zeebe/engine/state/ScheduledTaskDbState.java)).

**Comment by Zelldon** [comment]
Not sure whether @oleschoenburg is working on this? I think it was blocked by cleaning up the State changes in the checkers? :thinking:

**Comment by berkaycanbc** [comment]
ZPA Triage:

Closing this as this is a duplicate of #12308. Please re-open this if you think it is still necessary.

**Comment by Zelldon** [comment]
Disagree. This issue is clearly a bug :bug: which hasn't be resolved until the issue #12308 has been implemented. Furthermore, it says optionally which doesn't makes clear to me whether we will do it always (so I don't see whether this would fix our issue).

**Comment by korthout** [comment]
Finally found some time to discuss this with @Zelldon:
- 🐞 It is a bug
  - the scheduled tasks that still use the [`ProcessingDbState`](https://github.com/camunda/zeebe/blob/clients/go/v8.2.5/engine/src/main/java/io/camunda/zeebe/engine/Engine.java#L74-L79) (e.g. `DeploymentRedistributor`) instead of the [`ScheduledTaskDbState`](https://github.com/camunda/zeebe/blob/clients/go/v8.2.5/engine/src/main/java/io/camunda/zeebe/engine/Engine.java#L80) may run within the same transaction as the last processed command because these tasks may be scheduled before the Streaming Platform commits the transaction
- 😌 The bug is unlikely to surface as a problem
- 👷 However, scheduled tasks should not re-use the processing transaction

🏗️ The bug can be resolved by switching each scheduled task over to use the `ScheduledTaskDbState`. This state has a separate transaction context from the `ProcessingDbState`, and its transaction is not managed from outside. Instead, every database request is executed in a new transaction and committed once finished.

📅 In the future, we may run the scheduled tasks concurrently to each other. To prepare for this we may need to give each scheduled task its own instance of `ScheduledTaskDbState` (with each their own transaction context).

---

Lastly, I talked with @oleschoenburg about these steps. They are currently working on an issue ([Scheduled tasks should not block processing #12302](https://github.com/camunda/zeebe/issues/12302)) that would require switching each scheduled task over to `ScheduledTaskDbState` anyways. They're willing to also provide each scheduled task with their own instance of `ScheduledTaskDbState` while they're at it :heart:

**Comment by korthout** [comment]
I'm closing this as resolved since #13524 resolved #12308, which indirectly resolves this issue. I originally kept it open because @oleschoenburg and I wanted to backport #13524, but doing so costs more effort than it solves.

If anyone has issues with this on previous versions, please re-open this. It would be beneficial for us to understand the impact this bug had in that case.
